### PR TITLE
File autocomplete

### LINKED ISSUE #151: Feature proposal: Autocomplete with Open File
Will be good to have autocomplete when opening files inside the editor.

### PR DESCRIPTION
**Pull request by richrd**
Fixes #151 .

### CODE REVIEW

File: suplemon/prompt.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        x_offset = self.line_offset()
+        # Render the caption
+        self.window.addstr(pos[1], 0, self.caption)
+        # Render input
+        self.render_line_normal(line, pos, x_offset, max_len)
+
+    def handle_input(self, event):
+        """Handle special bindings for the prompt."""
+        name = event.key_name
+        if name in ["ctrl+c", "escape"]:
```
Comment (twolfson) [inline suplemon/prompt.py]: I'm not sure `Ctrl+C` is international, maybe this should detect a `KeyboardInterrupt` request?

Comment (richrd) [inline suplemon/prompt.py]: KeyboardInterrupts are converted into ctrl+c events in the input handler, so it shouldn't be a problem.

Comment (twolfson) [inline suplemon/prompt.py]: Maybe a constant should be used then (e.g. something defined in `suplemon`, `KeyboardInterrupt` itself)? A string seems brittle and not representative of the action by the user

Comment (richrd) [inline suplemon/prompt.py]: I don't quite understand what the benefit is of defining ctrl+c differently? I think all key_names in events should all be of the same format. That makes configuring easy etc.

Comment (twolfson) [inline suplemon/prompt.py]: That seems reasonable, maybe we should use constants for all keys too (e.g. `KEYS.ESCAPE`)? The benefit would be typos are easily avoided due to (1) autocompletion and (2) the program throwing an error due to an invalid key

Comment (richrd) [inline suplemon/prompt.py]: If such constants are used, maybe they should be defined in `key_mappings.py`. But then to be consistent all keys should probably have their own constants like you said. Not sure if the benefit's are worth it right now but if I find extra time I might do it. PR's are welcome too of course.

One thing that is probably more confusing is, that modifier keys can only be defined in a specific order in the user keymap. For example `ctrl+alt+up` will work but `alt+ctrl+up` won't. That should probably be fixed.

Comment (twolfson) [inline suplemon/prompt.py]: It looks like `curses` has a lot of definitions already. Maybe we don't need a `key_mappings.py` at all?

With respect to ordering, nice catch. I can think of 2 solutions:

- Use classes (e.g. `new Key(keycode, modifiers={})`
    - Prob not that efficient, requires extra logic for equality
- Add normalization to guarantee ordering of modifiers (e.g. always normalize to `shift+ctrl+alt+meta+key` so `alt+ctrl+up` becomes `ctrl+alt+up`)

Comment (richrd) [inline suplemon/prompt.py]: Well, the idea for `key_mappings.py` was to be able to use the same format for key config as Sublime Text does. Even though `curses` has lots of definitions they're a bit cryptic and they vary slightly between Python 2 and 3. Somehow it just feels natural to use the readable versions in the code too.

I think I'll go with your second solution regarding the modifier ordering. Seems the most straight forward way to do it.

Comment (twolfson) [inline suplemon/prompt.py]: I think the source of my reasoning is constants are more developer friendly so they can go inside internal code whereas external code (e.g. bindings) should be more non-developer friendly so maybe use validation for that

**Comment by richrd** [comment]
That's quite right, I do agree with constants being more robust and dev friendly. I'll try to look at maybe defining my own sane constants. The curses ones are just too obscure and they're not very well documented.

**Pull request merged**